### PR TITLE
Enum support in Reflection and Schema.

### PR DESCRIPTION
**Pull request by mzhurovich**
@dkorolev What do you think?

### CODE REVIEW

File: TypeSystem/Reflection/schema.h [hunk 1]
```diff
@@ -48,6 +48,8 @@ namespace reflection {
 
 CURRENT_STRUCT(TypeInfo) {
   CURRENT_FIELD(type_id, TypeID, TypeID::INVALID_TYPE);
+  // Used only for storing the name of `enum` types.
+  CURRENT_FIELD(name, std::string);
```
Comment (dkorolev) [inline TypeSystem/Reflection/schema.h]: `enum_name` ? :)

File: TypeSystem/enum.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    names_[std::type_index(typeid(T))] = name;
+  }
+
+  template <typename T>
+  ENABLE_IF<std::is_enum<T>::value, std::string> GetName() const {
+    const std::type_index type_index = std::type_index(typeid(T));
+    if (names_.count(type_index) != 0u) {
+      return names_.at(type_index);
+    } else {
+      // TODO: exception?
```
Comment (dkorolev) [inline TypeSystem/enum.h]: Just `std::cerr << ...` followed by `std::exit(-1);` would be perfect for my taste.

File: TypeSystem/enum.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#include "../Bricks/template/enable_if.h"
+#include "../Bricks/util/singleton.h"
+
+namespace current {
+namespace reflection {
+
+struct EnumNameSingletonImpl {
+  template <typename T>
+  ENABLE_IF<std::is_enum<T>::value> Register(const char* name) {
+    std::lock_guard<std::mutex> lock_guard(mutex_);
```
Comment (dkorolev) [inline TypeSystem/enum.h]: If we need this, do we need it in a getter too?

**Comment by dkorolev** [comment]
Wow, great!

Do `enum`-s support forward declarations of some sort? (Perhaps some `TemplateEnum<...>`?)

The only thing that separates this pull request from perfect is `CURRENT_REGISTER_ENUM()` -- could it be some `CURRENT_ENUM(foo, uint64_t) {}` or `CURRENT_ENUM(foo, uint64_t) { Bar = 1ull }` ?

**Comment by mzhurovich** [comment]
@dkorolev The biggest problem is that there is no forward declaration for `enum`-s :(((

**Comment by mzhurovich** [comment]
@dkorolev PTAL

**Pull request merged**